pilot/pkg/model/sidecar.go

```diff
@@ -0,0 +1,324 @@
+	// that these services need
+	for _, s := range out.allImportedServices {
+		out.allImportedDestinationRules[s.Hostname] = ps.DestinationRule(&dummyNode, s.Hostname)
+	}
+	meshGateway := map[string]bool{IstioMeshGateway: true}
+	out.allImportedVirtualServices = out.selectVirtualServices(ps.VirtualServices(&dummyNode, meshGateway))
+	return out
+}
+
+// Services returns the list of services imported across all egress listeners by this
+// Sidecar config
+func (sc *SidecarScope) Services() []*Service {
+	if sc == nil {
+		return nil
+	}
+
+	return sc.allImportedServices
+}
+
+// VirtualServices returns the list of virtual services imported by this Sidecar config
+// across all egress listeners
+func (sc *SidecarScope) VirtualServices() []Config {
+	if sc == nil {
+		return nil
+	}
+
+	return sc.allImportedVirtualServices
+}
+
+// DestinationRule returns the destination rule applicable for a given hostname
+// used by CDS code
+func (sc *SidecarScope) DestinationRule(hostname Hostname) *Config {
+	if sc == nil {
+		return nil
+	}
+
+	return sc.allImportedDestinationRules[hostname]
+}
+
+// Services returns the list of services imported by this egress listener
+func (ilw *IstioEgressListenerWrapper) Services() []*Service {
+	if ilw == nil {
+		return nil
+	}
+
+	return ilw.importedServices
+}
+
+func convertIstioListenerToWrapper(ps *PushContext, sidecarConfig *Config,
+	istioListener *networking.IstioEgressListener) *IstioEgressListenerWrapper {
+
+	out := &IstioEgressListenerWrapper{
+		IstioListener: istioListener,
+		importMap:     make(map[string]Hostname),
+	}
+
+	if istioListener.Port != nil {
+		out.ListenerPort = &Port{
+			Name:     istioListener.Port.Name,
+			Port:     int(istioListener.Port.Number),
+			Protocol: ParseProtocol(istioListener.Port.Protocol),
+		}
+	}
+
+	if istioListener.Hosts != nil {
+		for _, h := range istioListener.Hosts {
+			parts := strings.SplitN(h, "/", 2)
+			out.importMap[parts[0]] = Hostname(parts[1])
+		}
+	}
+
+	dummyNode := Proxy{
+		ConfigNamespace: sidecarConfig.Namespace,
+	}
+
+	out.importedServices = out.selectServices(ps.Services(&dummyNode))
+
+	return out
+}
+
+// GetEgressListenerForPort returns the egress listener corresponding to
+// the listener port or the catch all listener
+func (sc *SidecarScope) getEgressListenerForPort(port int) *IstioEgressListenerWrapper {
+	if sc == nil {
+		return nil
+	}
+
+	for _, e := range sc.EgressListeners {
+		if e.ListenerPort == nil || e.ListenerPort.Port == port {
+			return e
+		}
+	}
+	return nil
+}
+
+// selectVirtualServices returns the list of virtual services selected
+// across all egress listeners
+func (sc *SidecarScope) selectVirtualServices(configs []Config) []Config {
+
+	importedConfigs := make([]Config, 0)
+	for _, c := range configs {
+		configNamespace := c.Namespace
+		rule := c.Spec.(*networking.VirtualService)
+		for _, ilw := range sc.EgressListeners {
```

Comment:
It sounds like a better data structure is needed here - some map by namespace or similar. Double loop is usually a bad sign for perf. At least a TODO given the time pressure, but likely will be a hotspot.

Reply:
If you can show me how to a map will help in this selection procedure, it would be great, as I have no idea what you are talking about in this context. This function is called once per sidecar config. Not once per proxy